src/experimental/tools/videoThumbnailLoader/thumbnail_loader.ts

```diff
@@ -0,0 +1,349 @@
+ */
+function getSegmentData(
+  thumbnails: IThumbnailInfo[],
+  mediaElement: HTMLMediaElement
+): Observable<ArrayBuffer[]> {
+
+  const thumbnailsToLoad = thumbnails.filter((t) => {
+    const tRange = { start: t.start, end: t.start + t.duration };
+    const mediaRanges = convertToRanges(mediaElement.buffered);
+    return arrayFindIndex(mediaRanges, (mr) => {
+      return tRange.start >= mr.start && tRange.end <= mr.end;
+    }) === -1;
+  });
+
+  if (!thumbnailsToLoad.length) {
+    return EMPTY;
+  }
+
+  const loadedData$ = thumbnailsToLoad.map(({ mediaURL }) => {
+    return request({
+      url: mediaURL,
+      sendProgressEvents: false,
+      responseType: "arraybuffer",
+    });
+  });
+
+  return combineLatest(loadedData$).pipe(
+    map((loadedData) => {
+      const responseDatas = loadedData.map(({ value }) => value.responseData);
+      return responseDatas;
+    }),
+    take(1)
+  );
+}
+
+/**
+ * This tool, as a supplement to the RxPlayer, intent to help creating thumbnails
+ * from a video source.
+ *
+ * From a given video adaptation, the tools will extract a "thumbnail track",
+ * either from a trickMode track (whose light chunks are adapted from such use case)
+ * or direclty from the media content.
+ */
+export default class VideoThumbnailLoader {
+  private readonly _thumbnailVideoElement: HTMLVideoElement;
+  private readonly _bufferedDataRanges: Array<{
+    start: number;
+    end: number;
+  }>;
+
+  private _thumbnailTrack: IThumbnailTrack;
+
+  private _setTime$: Subject<number>;
+  private _setTimeSubscription$: Subscription;
+
+  constructor(
+    videoElement: HTMLVideoElement,
+    trickModeTrack: Representation
+  ) {
+    // readonly
+    this._thumbnailVideoElement = videoElement;
+    this._bufferedDataRanges = [];
+
+    // nullable
+    this._thumbnailTrack = this.updateThumbnailTrack(trickModeTrack);
+    this._setTime$ = new Subject();
+
+    const videoSourceInfos$ = prepareSourceBuffer(
+      this._thumbnailVideoElement,
+      this._thumbnailTrack.codec
+    ).pipe(
+      mergeMap(({ videoSourceBuffer }) => {
+        const { initURL: init, codec } = this._thumbnailTrack;
+        return request({ url: init,
+                         sendProgressEvents: false,
+                         responseType: "arraybuffer",
+        }).pipe(
+          mergeMap((e) => {
+            const { value: { responseData }} = e;
+            return videoSourceBuffer.appendBuffer({
+              initSegment : responseData,
+              segment: null,
+              codec,
+              timestampOffset: 0,
+            });
+          }),
+          mapTo(videoSourceBuffer)
+        );
+      }),
+      catchError(() => {
+        throw new Error("VideoThumbnailLoaderError: Couldn't open media source.");
+      }),
+      shareReplay()
+    );
+
+    this._setTimeSubscription$ = this._setTime$.pipe(
+      filter((time, i) => time !== this._thumbnailVideoElement.currentTime || i === 0),
```

Comment:
Weird condition (the first one). We should not rely on the current time but only on `buffered`, no?